### DISCUSSION TITLE
Scaling Postgres Prisma

### DISCUSSION THREAD

**Discussion by elie222** [prisma/prisma#9529] [Q&A]
Started using Prisma recently and running into scaling issues pretty early on.

We're listening to an API and storing data in the database every few seconds. Roughly 30 records per second but could easily go up. With MongoDB we were fine and didn't run into scaling issues but with PostgreSQL we've seen issues very early on.

The specific issues we've seen with Prisma:
 `Timed out fetching a new connection from the connection pool. (More info: http://pris.ly/d/connection-pool, Current connection limit: 17)`
 
 We've tried a bunch of different settings, using different connection timeouts, and pgbouncer but neither seem to help. Part of what we're doing uses serverless but not sure that is what's causing the issue as a non serverless script is hitting limits pretty quickly (starts okay but hits limits after a minute or so).
 
Our db is currently 1gb with 1 cpu. I can look to increase this further although it will start to get expensive quite quickly if we do this. Would love to be able to run at lower rates.

Any suggestions on how to handle these issues? Is there an automatic retry package with Prisma or I'd have to handle that myself?

I wouldn't mind handling these issues when we actually hit scale but atm feels like we're just getting started and it's strange we've already hit limits.

**Comment by ryands17** [comment]
@elie222 👋 

Could you share what parameters you have in the connection string? The current connection limit seems 17, but are you noticing a load on your database currently and can you see the no. of connections there?

**Reply by elie222** [reply]
I didn't have it set. But now setting it to 50 I get a similar error:
`?pool_timeout=10&connection_limit=50`

Error:

> Timed out fetching a new connection from the connection pool. (More info: http://pris.ly/d/connection-pool, Current connection limit: 50)
Error in upsertEvent due to Error: 
Invalid `prisma.event.upsert()` invocation:

**Reply by elie222** [reply]
TBH I labelled this post scaling Postgres but I don't even want that. I just don't want Prisma to try open 5000 connections when the limit is 50.

**Reply by ryands17** [reply]
I don't think it would open more connections than specified. How many connections does your database show?

**Reply by elie222** [reply]
Should be able to support 97 total. Here you can see a graph of what happened:
![image](https://user-images.githubusercontent.com/3090527/135462760-ec081b83-65e4-4779-8bf7-665b67c89736.png)

What we're doing is running a for loop that does a lot of async/await upserts.

And Prisma is trying to open more and more connections for each request. When I run the upserts serially and await each one things work fine but takes way longer. So ideally I just want to make as many upsert requests as possible and for Prisma to figure out when there are too many based on limits I set.

**Reply by elie222** [reply]
This database is hosted on render.com.

**Reply by ryands17** [reply]
Are you creating a new `PrismaClient` instance per request or is it being spawned somewhere?

**Reply by elie222** [reply]
All we're doing:

```ts
import { PrismaClient, Prisma } from "@prisma/client";

export const prisma = new PrismaClient();

export { Prisma };
```

And then usage:

```ts
    return await prisma.event.upsert({
      where: { eventId: event.eventId },
      create: event,
      update: event,
    })
```

**Reply by ryands17** [reply]
Would you mind sharing a reproduction of this behaviour? If `PrismaClient` is using more connections than defined in the limit, then it's a bug we would like to fix.

**Reply by elie222** [reply]
Created a repo that demos this error. See the repo README for instructions:
https://github.com/elie222/prisma-connection-bug

In the example I try to do 10k inserts and I get a connection error at some point.

**Reply by janpio** [reply]
Can we turn this into an issue so we can follow this via our normal process? That will also ask you the right questions about information we could use to properly reproduce this.

And thanks for doing this by the way!

**Reply by elie222** [reply]
Posted here: https://github.com/prisma/prisma/issues/9569

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
